http/handler/dockerhub/dockerhub_status.go

```diff
@@ -0,0 +1,140 @@
+package dockerhub
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	httperror "github.com/portainer/libhttp/error"
+	"github.com/portainer/libhttp/request"
+	"github.com/portainer/libhttp/response"
+)
+
+type dockerhubStatusPayload struct {
+	Authentication bool   `json:"authentication"`
+	Username       string `json:"username"`
+	Password       string `json:"password"`
+}
+
+func (payload *dockerhubStatusPayload) Validate(r *http.Request) error {
+	if payload.Authentication {
+		if payload.Username == "" || payload.Password == "" {
+			return errors.New("Missing username or password")
+		}
+	}
+
+	return nil
+}
+
+type dockerhubStatusResponse struct {
+	Remaining int `json:"remaining"`
+	Limit     int `json:"limit"`
+}
+
+// GET request on /api/endpoints/{id}/dockerhub/status
+func (handler *Handler) dockerhubStatus(w http.ResponseWriter, r *http.Request) *httperror.HandlerError {
+	var payload dockerhubStatusPayload
+	err := request.DecodeAndValidateJSONPayload(r, &payload)
+	if err != nil {
+		return &httperror.HandlerError{http.StatusBadRequest, "Invalid request payload", err}
+	}
+
+	httpClient := &http.Client{
+		Timeout: time.Second * 3,
+	}
+	token, err := getDockerHubToken(httpClient, &payload)
+	if err != nil {
+		return &httperror.HandlerError{http.StatusInternalServerError, "Unable to retrieve DockerHub token from DockerHub", err}
+	}
+
+	resp, err := getDockerHubLimits(httpClient, token)
+	if err != nil {
+		return &httperror.HandlerError{http.StatusInternalServerError, "Unable to retrieve DockerHub rate limits from DockerHub", err}
+	}
+
+	return response.JSON(w, resp)
+}
+
+func getDockerHubToken(httpClient *http.Client, dockerhub *dockerhubStatusPayload) (string, error) {
+	type dockerhubTokenResponse struct {
+		Token string `json:"token"`
+	}
+
+	requestURL := "https://auth.docker.io/token?service=registry.docker.io&scope=repository:ratelimitpreview/test:pull"
```

Comment:
as for portainer we should use some constant

Reply:
this code should be in a library :(